src/common/core/metaquerycore.js

```diff
@@ -0,0 +1,261 @@
+                        if (innerCore.isTypeOf(validNodes[i], metaNodes[keys[j]])) {
+                            inAspect = true;
+                            break;
+                        }
+                    }
+                    if (!inAspect) {
+                        validNodes.splice(i, 1);
+                    }
+                }
+            }
+            return validNodes;
+        };
+
+        this.getValidSetElementsMetaNodes = function (parameters) {
+            var validNodes = [],
+                node = parameters.node,
+                metaNodes = self.getAllMetaNodes(node),
+                keys = Object.keys(metaNodes || {}),
+                i, j,
+                typeCounters = {},
+                members = parameters.members || [],
+                rules = self.getPointerMeta(node, parameters.name) || {},
+                temp;
+
+            for (i = 0; i < keys.length; i += 1) {
+                temp = metaNodes[keys[i]];
+                while (temp) {
+                    if (rules[innerCore.getPath(temp)]) {
+                        validNodes.push(metaNodes[keys[i]]);
+                        break;
+                    }
+                    temp = innerCore.getBase(temp);
+                }
+            }
+
+            //before every next step we check if we still have potential nodes
+            if (validNodes.length === 0) {
+                return validNodes;
+            }
+
+            if (parameters.sensitive === true) {
+                sensitiveFilter(validNodes);
+            }
+
+            //before every next step we check if we still have potential nodes
+            if (validNodes.length === 0) {
+                return validNodes;
+            }
+
+            if (parameters.multiplicity === true) {
+                if (rules.max && rules.max > -1 && innerCore.getMemberPaths(node).length >= rules.max) {
+                    validNodes = [];
+                    return validNodes;
+                }
+
+                if (members.length === 0) {
+                    return validNodes; //we cannot check type-multiplicity without children
+                }
+
+                delete rules.max;
+                delete rules.min;
+
+                //we need to clear nodes that are not on the meta sheet
+                // and we have to initialize the counters
+                keys = Object.keys(rules);
+                for (i = 0; i < keys.length; i += 1) {
+                    if (!metaNodes[keys[i]]) {
+                        delete rules[keys[i]];
+                    } else {
+                        typeCounters[keys[i]] = 0;
+                    }
+                }
+
+                keys = Object.keys(rules);
+                for (i = 0; i < members.length; i += 1) {
+                    for (j = 0; j < keys.length; j += 1) {
+                        if (innerCore.isTypeOf(members[i], metaNodes[keys[j]])) {
+                            typeCounters[keys[j]] += 1;
```

Comment:
This is potentially slow, make a FIXME note, or fix it.
http://jsperf.com/array-push-vs-splice